controllers/nutanixfailuredomain_controller.go

```diff
@@ -0,0 +1,231 @@
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=nutanixfailuredomains/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=nutanixfailuredomains/finalizers,verbs=get;update;patch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the NutanixFailureDomain object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (r *NutanixFailureDomainReconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, reterr error) {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Reconciling the NutanixFailureDomain")
+
+	// Fetch the NutanixFailureDomain instance
+	fd := &infrav1.NutanixFailureDomain{}
+	if err := r.Get(ctx, req.NamespacedName, fd); err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			log.Info("The NutanixFailureDomain object not found. Ignoring since object must be deleted.")
+			return reconcile.Result{}, nil
+		}
+
+		// Error reading the object - requeue the request.
+		log.Error(err, "failed to fetch the NutanixFailureDomain object")
+		return reconcile.Result{}, err
+	}
+
+	// Initialize the patch helper.
+	patchHelper, err := patch.NewHelper(fd, r.Client)
+	if err != nil {
+		log.Error(err, "Failed to configure the patch helper")
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	defer func() {
+		// Always attempt to Patch the NutanixFailureDomain object and its status after each reconciliation.
+		if err := patchHelper.Patch(ctx, fd); err != nil {
+			reterr = kerrors.NewAggregate([]error{reterr, err})
+			log.Error(reterr, "Failed to patch NutanixFailureDomain.")
+		} else {
+			log.Info("Patched NutanixFailureDomain.", "status", fd.Status, "finalizers", fd.Finalizers)
+		}
+	}()
+
+	// Add finalizer first if not set yet
+	if !ctrlutil.ContainsFinalizer(fd, infrav1.NutanixFailureDomainFinalizer) {
+		if ctrlutil.AddFinalizer(fd, infrav1.NutanixFailureDomainFinalizer) {
+			// Add finalizer first avoid the race condition between init and delete.
+			log.Info("Added the finalizer to the object", "finalizers", fd.Finalizers)
+			return reconcile.Result{}, nil
+		}
+	}
+
+	// Handle deletion
+	if !fd.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, fd)
+	}
+
+	return r.reconcileNormal(ctx, fd)
+}
+
+func (r *NutanixFailureDomainReconciler) reconcileDelete(ctx context.Context, fd *infrav1.NutanixFailureDomain) (ctrl.Result, error) {
```

Comment:
Currently the reconcileDelete only looks at whether existing NutanixMachines are placed on a failure domain. If not, it removes the finalizer. In theory, this behavior should also extend to checking failure domains referenced in NutanixCluster.spec.ControlPlaneFailureDomains.

At the moment, the NutanixClusterController would remove the failure domain from the cluster's status to ensure a new control plane machine doesn't get created with a deleted failure domain and mark the NutanixCluster with a `FailureDomainsMisconfigured` condition.

We discussed offline on this a bit and decided not to hold the PR on this issue. However, I will experiment with some more scenarios and try to work out whether or not this should be the desired behavior.